### PR TITLE
feat: add `EngineApi` extension trait

### LINKED ISSUE #658: [Feature] Add engine api extension
### Component

provider, pubsub, rpc

### Describe the feature you would like

mostly for testing purposes, it would be nice to have an `EngineApiExt` trait similar to 

https://github.com/alloy-rs/alloy/blob/c058e32d2025a2fd60b2617554ade7afeaca9c47/crates/provider/src/ext/admin.rs#L11

see also:

https://github.com/paradigmxyz/reth/blob/0938504f4a265eec74e0b833ab5c41d66649bdee/crates/rpc/rpc-api/src/engine.rs#L27-L27

## TODO 
* add extension trait for engine API
* can use ethereum engine types, but I think eventually we'd need a helper trait because some request params are also network specific
* should be feature gated because not useful for regular usage

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by leruaa**
Closes  #658

### CODE REVIEW

**Review by onbjerg — CHANGES_REQUESTED** [review]
this is missing https://github.com/ethereum/execution-apis/blob/main/src/engine/identification.md

File: crates/provider/Cargo.toml [hunk 1]
```diff
@@ -70,3 +71,4 @@ reqwest-default-tls = ["alloy-transport-http?/reqwest-default-tls"]
 reqwest-rustls-tls = ["alloy-transport-http?/reqwest-rustls-tls"]
 reqwest-native-tls = ["alloy-transport-http?/reqwest-native-tls"]
 anvil = ["reqwest", "dep:alloy-node-bindings", "dep:alloy-signer-wallet"]
+engine = ["dep:alloy-rpc-types-engine"]
```
Comment (mattsse) [inline crates/provider/Cargo.toml:74]:
```suggestion
engine-api = ["dep:alloy-rpc-types-engine"]
```

File: crates/provider/src/ext/engine.rs [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+    PayloadStatus,
+};
+use alloy_transport::{Transport, TransportResult};
+
+use crate::Provider;
+
+/// Extension trait that gives access to engine api RPC methods.
+#[cfg_attr(target_arch = "wasm32", async_trait::async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait::async_trait)]
+pub trait EngineApi<N, T>: Send + Sync {
```
Comment (mattsse) [inline crates/provider/src/ext/engine.rs:15]: this should make a note that the provider should use  JWT authentication layer

**Review by mattsse — CHANGES_REQUESTED** [review]
nice, smol nit

File: crates/provider/src/ext/engine.rs [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+    /// See also <https://github.com/ethereum/execution-apis/blob/6452a6b194d7db269bf1dbd087a267251d3cc7f8/src/engine/shanghai.md#engine_getpayloadbodiesbyrangev1>
+    async fn get_payload_bodies_by_range_v1(
+        &self,
+        start: u64,
+        count: u64,
+    ) -> TransportResult<ExecutionPayloadBodiesV1>;
+
+    /// Returns the execution client version information.
+    ///
+    /// See also <https://github.com/ethereum/execution-apis/blob/main/src/engine/identification.md#engine_getclientversionv1>
```
Comment (onbjerg) [inline crates/provider/src/ext/engine.rs:135]: let's add a small doc explaining that the param is the CL client version

**Review by onbjerg — CHANGES_REQUESTED** [review]
otherwise lgtm

**Review by mattsse — APPROVED** [review]

**Pull request merged**